csharp/src/Google.Protobuf/JsonToken.cs

```diff
@@ -39,7 +39,7 @@ internal static JsonToken Value(double value)
 
         internal enum TokenType
         {
-            Null,
+            Null = 1, //zero is reserved for check with default(JsonToken)
```

Comment:
@jskeet As far as I know - structs cannot be null, you can only set them to default(struct). If you check Equal method will se that it checks for every property and in case of JsonToken.Null instance cannot be distinguished from defaut(JsonToken). With this change default(JsonToken) will have 0 in type property and will differ.

I will check and fix other comments + will add benchmarks for this PR and for both.

Reply:
> As far as I know - structs cannot be null

Except for nullable variables - that's what I was suggested; that we change `bufferedToken` to be a `JsonToken?` instead of `JsonToken`.

Reply:
Right, my benchmark showed the StringBuilder change making a big difference - but not the JsonToken one. I'll look at that again next week with BenchmarkDotnet.